docs/electrical/embedded/styleguide.md

```diff
@@ -1,3 +1,136 @@
+
+??? info "Revision History"
+    - *Revision 0.1 by William K, 11 Jun 2023*
+    - *Revision 0.2 by William K, 12 Aug 2023*
+
+---
+
+## Rules
+
+### General Rules
+#### Line Width
+1. No line width is enforced as we do not do paper code reviews. Please use your best descretion for when a line is considered too long.
+2. The preprocessor directive (```#define```) shall not be used to alter or rename any keyword or other part of the C programming language.
+3. The minimum target standard for both C and C++ is **C/C++ 11**. However, if newer versions such as 14 or 17 are available for the toolchain in use, they are preferred.
+   
+#### Braces
+1. Left braces should appear on the same line as the statement they are connected with. Right braces should always be on the line after the line of code in the block if the block is more than one line.
+   
+    Should the block be only one line long for a *conditional*, the entire statement including the right brace should be placed on the same line
+    ``` c title="Example"
+    if (condition) {code block;}
+    else {
+        code block line 1;
+        code block line 2;
+        code block line 3;
+    }
+    ```
+
+#### Parentheses
+1. To maintain clarity of long expressions, parentheses should be used to ensure proper execution order and provide better information to maintainers.
+2. For further clarity, each side of a logical operator should be surrounded with parentheses, unless it is a single level operation.
+```c title="Example"
+((current_depth == target_depth) && (current_depth < pool_depth))
+
+(soft_arm && hard_arm)
+```
+
+#### Common Abreviations
+1. Unless it is an abbreviation in common use and understood widely among programmers, refrain from using abreviations as they can obfuscate the meaning of code.
+2. Common variable names, such as ```i``` for a loop counter can be used under the condition that it is of limited scope and purpose, and that the context it is in is exceedingly clear.
+
+#### Casts
+1. Should a cast be required, the relevant lines should be commented to explain assumptions made.
+
+#### Keywords
+1. The use of ```goto``` should be exceedingly limited to only cases where it is justifiably required, and no other solution can be found (including rewriting other code to make the ```goto``` unecessary).
+2. Uses of the ```continue``` keyword should be limited, but are allowed where it simplifies the code structure and makes it more legible.
+3. *The ```static``` keyword needs some though put towards its use cases, I don't know if anything needs to be defined for this.*
+4. ```const``` should be applied in all locations where appropriate. In some cases, it may be more applicable than a preprocessor directive, while the preprocessor directive may be more applicable in others.
+5. ```volatile``` should be used in all cases where a variable can be modified by anything outside of the normal control flow of the code (i.e. an ISR).
```

Comment:
Important for variables shared between RTOS threads too. Often overlooked, so I'd note it explicitly.